crates/exex/src/manager.rs

```diff
@@ -0,0 +1,446 @@
+        &self,
+        notification: CanonStateNotification,
+    ) -> Result<(), SendError<CanonStateNotification>> {
+        self.exex_tx.send(notification)
+    }
+
+    /// Asynchronously send a notification over the channel to all execution extensions.
+    ///
+    /// The returned future resolves when the notification has been delivered. If there is no
+    /// capacity in the channel, the future will wait.
+    pub async fn send_async(
+        &mut self,
+        notification: CanonStateNotification,
+    ) -> Result<(), SendError<CanonStateNotification>> {
+        self.ready().await;
+        self.exex_tx.send(notification)
+    }
+
+    /// Get the current capacity of the ExEx manager's internal notification buffer.
+    pub fn capacity(&self) -> usize {
+        self.current_capacity.load(Ordering::Relaxed)
+    }
+
+    /// Whether there is capacity in the ExEx manager's internal notification buffer.
+    ///
+    /// If this returns `false`, the owner of the handle should **NOT** send new notifications over
+    /// the channel until the manager is ready again, as this can lead to unbounded memory growth.
+    pub fn has_capacity(&self) -> bool {
+        self.current_capacity.load(Ordering::Relaxed) > 0
+    }
+
+    /// Returns `true` if there are ExEx's installed in the node.
+    pub fn has_exexs(&self) -> bool {
+        self.num_exexs > 0
+    }
+
+    /// The finished height of all ExEx's.
+    ///
+    /// This is the lowest common denominator between all ExEx's. If an ExEx has not emitted a
+    /// `FinishedHeight` event, it will be `None`.
+    ///
+    /// This block is used to (amongst other things) determine what blocks are safe to prune.
+    ///
+    /// The number is inclusive, i.e. all blocks `<= finished_height` are safe to prune.
+    pub fn finished_height(&mut self) -> Option<BlockNumber> {
+        *self.finished_height.borrow_and_update()
+    }
+
+    /// Wait until the manager is ready for new notifications.
+    pub async fn ready(&mut self) {
+        poll_fn(|cx| self.poll_ready(cx)).await
+    }
+
+    /// Wait until the manager is ready for new notifications.
+    pub fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        // if this returns `Poll::Ready(None)` the stream is exhausted, which means the underlying
+        // channel is closed.
+        //
+        // this can only happen if the manager died, and the node is shutting down, so we ignore it
+        let mut pinned = std::pin::pin!(&mut self.is_ready);
+        if pinned.poll_next_unpin(cx) == Poll::Ready(Some(true)) {
+            Poll::Ready(())
+        } else {
+            Poll::Pending
+        }
+    }
+}
+
+impl Clone for ExExManagerHandle {
+    fn clone(&self) -> Self {
+        Self {
+            exex_tx: self.exex_tx.clone(),
+            num_exexs: self.num_exexs,
+            is_ready_receiver: self.is_ready_receiver.clone(),
+            is_ready: WatchStream::new(self.is_ready_receiver.clone()),
```

Comment:
you can't clone a `WatchStream` so we need to keep the receiver as an additional field and create a new `WatchStream` on clone